File: content/step0/tree/tools/resolve_ref.py
from sys import argv


def resolve_ref(git_ref, directory):
    if git_ref == 'HEAD':
        with open('/'.join([directory, git_ref])) as f:
            pointed_ref = f.read()[5:-1]  # Remove trailing newline
            return resolve_ref(pointed_ref, directory)
    else:
        with open('/'.join([directory, git_ref])) as f:
            return f.read()[:-1]

if __name__ == "__main__":
    ref = argv[1]
    print(resolve_ref(ref, ".git"))

File: content/step0/tree/tools/ResolveRef.hs
#!/usr/bin/env stack
-- stack --resolver lts-9.0 script

module Main where

import System.Environment
import System.FilePath

resolveRef :: String -> FilePath -> IO String
resolveRef ref dir = case ref of
    "HEAD" -> do
        content <- readFile $ dir </> "HEAD"
        let pointedRef = init . drop 5 $ content
        resolveRef pointedRef dir
    _      -> do
        content <- readFile $ dir </> ref
        return $ init content

main :: IO ()
main = do
    (ref:_) <- getArgs
    resolved <- resolveRef ref ".git"
    putStrLn resolved
